client/devdocs/design/search-collection.jsx

```diff
@@ -0,0 +1,97 @@
+/**
+* External dependencies
+*/
+import React from 'react';
+
+/**
+ * Internal dependencies
+ */
+let FilterSummary, Hider;
+
+Hider = React.createClass( {
+	displayName: 'Hider',
+
+	propTypes: {
+		hide: React.PropTypes.bool,
+	},
+
+	shouldComponentUpdate( nextProps ) {
+		return this.props.hide !== nextProps.hide;
+	},
+
+	render() {
+		return (
+			<div style={ this.props.hide ? { display: 'none' } : {} }>
+				{ this.props.children }
+			</div>
+		);
+	}
+} );
+
+FilterSummary = React.createClass( {
+	render() {
+		var names;
+
+		if ( this.props.items.length === 0 ) {
+			return ( <p>No matches found</p> );
+		} else if ( this.props.items.length === this.props.total || this.props.items.length === 1 ) {
+			return null;
+		}
+
+		names = this.props.items.map( function( item ) {
+			return item.props.children.type.displayName;
+		} );
+
+		return (
+			<p>Showing: { names.join( ', ' ) }</p>
+		);
+	}
+} );
+
+export default React.createClass( {
+	displayName: 'Collection',
+
+	shouldWeHide: function( example ) {
+		var filter, searchString;
+
+		filter = this.props.filter || '';
+		searchString = example.type.displayName;
+
+		if ( this.props.component ) {
+			return example.type.displayName.toLowerCase() !== this.props.component.replace( /-([a-z])/g, '$1' );
+		}
+
+		if ( example.props.searchKeywords ) {
+			searchString += ' ' + example.props.searchKeywords;
+		}
+
+		return ! ( ! filter || searchString.toLowerCase().indexOf( filter ) > -1 );
+	},
+
+	visibleExamples: function( examples ) {
+		return examples.filter( ( child ) => {
+			return !child.props.hide;
+		} );
+	},
+
+	render: function() {
+		var summary, examples;
+
+		examples = this.props.children.map( ( example ) => {
+			return (
+				<Hider hide={ this.shouldWeHide( example ) } key={ 'example-' + example.type.displayName }>
+					{ example }
+				</Hider>
+			);
+		} );
+
+		summary = !this.props.component ? <FilterSummary items={ this.visibleExamples( examples ) } total={ this.props.children.length } /> : null;
```

Comment:
Optional, but since you're already converting some ES6, we could do away with the `var` altogether (discouraged under ES6 syntax). A smaller point, but ternaries like these are pretty hard to interpret, and would be much more legible if rewritten as:

```jsx
let summary;
if ( ! this.props.component ) {
	summary = (
		<FilterSummary
			items={ this.visibleExamples( examples ) }
			total={ this.props.children.length } />
	);
}
```

Noting:
- Space after the negation (not captured by linter, but [documented guideline](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#spacing))
- A rendered variable as a child can be left `undefined`, doesn't need to be explicitly `null`
- Line length should usually not exceed 80 characters

(Understandably, many of these are remnants of existing code, so not on you specifically)